Sources_v3/Utils/Internet Connection/InternetConnection.swift

```diff
@@ -0,0 +1,250 @@
+        case expensive
+        
+        /// Internet connection uses Low Data Mode.
+        /// Recommendations for Low Data Mode: don't autoplay video, music (high-quality) or gifs (big files).
+        /// Supports only by iOS 13+
+        case constrained
+    }
+}
+
+// MARK: - Internet Connection Monitor
+
+private extension InternetConnection {
+    /// The default Internet connection monitor for iOS 12+.
+    /// It uses Apple Network API.
+    @available(iOS 12, *)
+    class Monitor: InternetConnectionMonitor {
+        private var monitor: NWPathMonitor?
+        
+        weak var delegate: InternetConnectionDelegate?
+        
+        var status: InternetConnection.Status {
+            if let path = monitor?.currentPath {
+                return status(from: path)
+            }
+            
+            return .unknown
+        }
+        
+        func start() {
+            guard monitor == nil else { return }
+            
+            monitor = createMonitor()
+            monitor?.start(queue: .global())
+        }
+        
+        func stop() {
+            monitor?.cancel()
+            monitor = nil
+            delegate?.internetConnectionStatusDidChange(status: .unknown)
+        }
+        
+        private func createMonitor() -> NWPathMonitor {
+            let monitor = NWPathMonitor()
+            monitor.pathUpdateHandler = { [unowned self] in self.updateStatus(with: $0) }
+            return monitor
+        }
+        
+        private func updateStatus(with path: NWPath) {
+            log.info("Internet Connection info: \(path.debugDescription)")
+            delegate?.internetConnectionStatusDidChange(status: status(from: path))
+        }
+        
+        private func status(from path: NWPath) -> InternetConnection.Status {
+            guard path.status == .satisfied else {
+                return .unavailable
+            }
+            
+            let quality: InternetConnection.Quality
+            
+            if #available(iOS 13.0, *) {
+                quality = path.isConstrained ? .constrained : (path.isExpensive ? .expensive : .great)
+            } else {
+                quality = path.isExpensive ? .expensive : .great
+            }
+            
+            return .available(quality)
+        }
+    }
+}
+
+// MARK: Legacy Internet Connection Monitor for iOS 11 only
+
+private extension InternetConnection {
+    class LegacyMonitor: InternetConnectionMonitor {
+        /// A Reachability instance for Internet connection monitoring.
+        private lazy var reachability = createReachability()
+        
+        weak var delegate: InternetConnectionDelegate?
+        
+        var status: InternetConnection.Status {
+            if let reachability = reachability {
+                return status(from: reachability)
+            }
+            
+            return .unknown
+        }
+        
+        func start() {
+            DispatchQueue.main.async {
```

Comment:
Why `.main`?

Reply:
I moved this implementation from `v2`. I can't say why. but probably it was needed.